pricegraph/src/orderbook.rs

```diff
@@ -149,12 +162,88 @@ impl Orderbook {
         self.projection.find_edge(buy, sell)
     }
 
-    /// Retrieve the weight of an edge in the projection graph. This is used for
-    /// testing that the projection graph is in sync with the order map.
-    #[cfg(test)]
-    fn get_projected_pair_weight(&self, pair: TokenPair) -> f64 {
-        let edge = self.get_pair_edge(pair).unwrap();
-        self.projection[edge]
+    /// Fills a trading path through the orderbook to maximum capacity, reducing
+    /// the remaining order amounts and user balances along the way, returning
+    /// the amount of flow that left the first node in the path or `None` if the
+    /// path was invalid.
+    ///
+    /// Note that currently, user buy token balances are not incremented as a
+    /// result of filling orders along a path.
+    pub fn fill_path(&mut self, path: &[TokenId]) -> Option<f64> {
+        let capacity = self.find_maximum_capacity(path)?;
+        self.fill_path_with_capacity(path, capacity)
+            .expect("failed to fill detected path with capacity");
+
+        Some(capacity)
+    }
+
+    /// Gets the maximum capacity of a path expressed in an amount of the
+    /// starting token. Returns `None` if the path doesn't exist.
+    fn find_maximum_capacity(&self, path: &[TokenId]) -> Option<f64> {
+        let mut capacity = f64::INFINITY;
+        let mut transient_price = 1.0;
+        for pair in pairs_on_path(path) {
+            let order = self.orders.best_order_for_pair(pair)?;
+            transient_price *= order.price;
+
+            let sell_amount = order.get_effective_amount(&self.users);
+            capacity = num::min(capacity, sell_amount * transient_price);
```

Comment:
Why is the sell_amount (in some token along the path) multiplied with the transient_price, which as I understand includes information about all previous edges already. 

I would think, capacity is related to the amount at the current end of the path and also already incorporates all information about the previous prices.

Thus I'd expect

capacity = num::min(capacity, sell_amount) * order.price;

Reply:
We are trying to find the capacity of the path expressed in the starting token, this makes things much easier when doing successive shortest path as we want to find enough capacity over `n` paths to completely drain an amount.

Reply:
Right, this makes perfect sense! But don't we have to divide sellAmount with transient price in this case?

E.g. Take the order to "buy at least 1A for at most 2B" --> price 1/2 --> transient_price = 1* 1/2 = 1/2

Now let's say we have an order selling 1A. In order to convert that sell amount into starting token (B) we would have to divide by the transient price not multiply, wouldn't we?

Reply:
I had price graph nightmares from thinking about this so much...

So the order "buy at least 1A for at most 2B" does indeed have a price of 1/2, but in our price graph, since we are pushing tokens through the graph, the graph direction is from the buy token to the sell token, since we can only offer token A to a funded order that is willing to buy it. In this situation, the sell amount is 2, so the maximum amount of token A we can send between the `A->B` order is 1 (`(sell_amount = 2) * (transient_price = 0.5)`).

Reply:
What I meant by:
> the graph direction is from the buy token to the sell token, since we can only offer token A to a funded order that is willing to buy it

Is that if we are trying to answer the question "what is the best value for `x` for my order: buy 1A for at most `x`B", then we do so by examining all orders that are "buy at most `b`B for at least `a`A" so that we can "push" our B token to A.

Given this, as you said, the best order for us is a small as possible `x` value (so the price goes up), and in order to achieve that, we need minimize the price of the other orders (hence the smaller price is better for getting a lower cost pushing some token through the graph) which is to say, minimize `a/b` (so higher `b` is better for this point of view).

Reply:
I will try to illustrate this with an example, if we want to find the best possible `x` for the order:
- "buy at least `x`A for at most 2B"

Given the orders:
- "buy at least 1B for 2A"
- "buy at least 1B for 3C"
- "buy at least 6C for 3A"

This means, in order to fill your order A->B, you can take the path B->A or B->C->A.

1. For the first path, the transient price is 1/2 (1B / 2A), and so the amount of B you can send through that path is `((sell_amount = 2A) * (1B / 2A)) = 1B`. This means you still have 1B of volume to push through the graph.
2. For the second path, there are two pairs:
  1. The first segment has a transient price of `1B / 3C`, and a capacity of 1B
  2. The second segment has a transient price of `1B * 6C / 3C * 3A = 2B / 3A` and a capacity of `((sell_amount = 3A) * (2B / 3A)) = 2B`
  3. This makes the final transient price be `2B / 3A` and the total capacity to be `min(1B, 2B) = 1B`.

I hope this helps explain it a bit.